lib/ocpp/v201/monitoring_updater.cpp

```diff
@@ -266,280 +279,349 @@ void MonitoringUpdater::on_variable_changed(const std::unordered_map<int64_t, Va
+        process_monitor_meta_internal(updater_monitor_meta);
 
-        if (it->second.csms_sent == false && it->second.cleared == true) {
-            // Based on spec comments, if a variable returns to normal and we're offline and
-            // it was not sent to the CSMS it can safely be discarded:
-            // "Only those monitoring events that were triggered & cleared during the offline period will remain
-            // invisible to CSMS."
-            should_clear = true;
-        } else if (it->second.csms_sent && it->second.cleared == true) {
-            // Also discard monitors that were sent to the CSMS and are cleared,
-            // that is, in their proper state
-            should_clear = true;
+        // If we are not offline, send the queued events generated by this meta
+        if (!is_offline) {
+            if (!updater_monitor_meta.generated_monitor_events.empty()) {
+                EVLOG_debug << "Sent data for monitor: " << updater_monitor_meta.monitor_meta.monitor;
+
+                // Send the events
+                notify_csms_events(updater_monitor_meta.generated_monitor_events);
+                updater_monitor_meta.generated_monitor_events.clear();
+
+                if (updater_monitor_meta.type == UpdateMonitorMetaType::TRIGGER) {
+                    // If we have a trigger mark the events as being sent
+                    // for the curent state
+                    updater_monitor_meta.meta_trigger.is_csms_sent = true;
+
+                    // If this was a state trigger, them also mark that
+                    // we sent this 'dangerous' state to the CSMS at least once
+                    // since in that case the clear logic changes
+                    if (updater_monitor_meta.meta_trigger.is_cleared == false) {
+                        updater_monitor_meta.meta_trigger.is_csms_sent_triggered = true;
+                    }
+                }
+            }
+        } else {
+            // If we are offline but we passed the 'should_process' test, it means that
+            // we should keep the generated events and send them at a further occasion
+            EVLOG_debug << "We are offline, cached generated events for later!";
         }
 
-        if (should_clear) {
-            EVLOG_debug << "Erased triggered monitor: [" << it->second.component << ":" << it->second.variable
-                        << "] since it was either cleared or it was sent to the CSMS and cleared";
-            it = triggered_monitors.erase(it);
+        if (should_remove_monitor_meta_internal(updater_monitor_meta)) {
+            it = updater_monitors_meta.erase(it);
         } else {
             ++it;
         }
     }
 }
 
+bool MonitoringUpdater::is_monitoring_enabled() {
+    return this->device_model->get_optional_value<bool>(ControllerComponentVariables::MonitoringCtrlrEnabled)
+        .value_or(false);
+}
+
+void MonitoringUpdater::get_monitoring_info(bool& out_is_offline, int& out_offline_severity,
```

Comment:
I was thinking that if we're using the out_params and that it might grow in the future, to respect that pattern, but there's no hard-requirement on that. Do you think it looks better returning the MonitoringBaseEnum?

Reply:
ok